### PR TITLE
2019 a11y fixes

### PR DESCRIPTION
**Pull request by zainafzal08**
This PR

- Fixes the hidden rad/deg UI to be visible and focusable.
- Refactors the Tab index for better screen reader support
- Added better speech labels to the different sections of the app
- Fixed a bug where the arrow key navigation would fail on not en languages.
- Updated color scheme of the calc keypad for better contrast against the focus ripple.
- Better handling of long numbers

Will squash before sending it off to the actual foam repo

### CODE REVIEW

**Comment by E-T-** [comment]
Please update the patch description to say what's changing in the description. Is it just colors? I know the commit messages kinda give this as well but I think there might be some things that can be evinced more easily from the description rather than through the log. Your call here.

**Comment by zainafzal08** [comment]
Closing this PR and opening a series of smaller more topic focused ones.

**Comment by zainafzal08** [comment]
Attempted to split this up into multiple PRs but many of the commits are sadly interdependent and it was causing more issues then it was solving, going to reopen this one and just provide more context

File: js/foam/apps/calc/Calc.js [hunk 1]
```diff
@@ -49,6 +49,21 @@ CLASS({
       name: 'CalcName',
       value: 'Calculator',
       translationHint: 'name of application for performing simple calculations'
+    },
```
Comment (zainafzal08) [inline js/foam/apps/calc/Calc.js]: This is how Foam lets us specify i18n strings, these are automatically translated if the language is non-eu

File: js/foam/apps/calc/Calc.js [hunk 2]
```diff
@@ -175,7 +190,7 @@ CLASS({
       name: "mult",
       label: "×",
       speechLabel: "multiply",
-      keyboardShortcuts: [ "*" ],
+      keyboardShortcuts: [ "*", "x" ],
```
Comment (zainafzal08) [inline js/foam/apps/calc/Calc.js]: Allows users to use 'x' to specify the multiply command, as per request of a11y folks

File: js/foam/apps/calc/Calc.js [hunk 3]
```diff
@@ -376,13 +391,23 @@ CLASS({
       name: 'deg',
       speechLabel: 'switch to degrees',
       translationHint: 'short form for "degrees" calculator mode',
-      code: function() { this.degreesMode = true; }
+      code: function() {
```
Comment (zainafzal08) [inline js/foam/apps/calc/Calc.js]: This code updates the label for the rad/deg label to match the state so chromevox can read the correct label.

File: js/foam/apps/calc/CalcView.js [hunk 4]
```diff
@@ -229,8 +233,8 @@ CLASS({
       font-size: 22px;
       font-weight: 400;
       opacity: 0;
-      padding-left: 8px;
-      padding-right: 10px;
+      margin-left: 8px;
```
Comment (zainafzal08) [inline js/foam/apps/calc/CalcView.js:255]: This is to fix how the calculator works on overflow

File: js/foam/apps/calc/CalcView.js [hunk 5]
```diff
@@ -403,6 +407,16 @@ CLASS({
     .inner-calc-display:focus .f1 {
       margin-left: calc(-13pt - 4px);
     }
+
```
Comment (zainafzal08) [inline js/foam/apps/calc/CalcView.js:429]: This gives the rad/deg label a focus UI

File: js/foam/apps/calc/SecondaryButtonsView.js [hunk 6]
```diff
@@ -29,28 +29,28 @@ CLASS({
           <div id="%%id" class="buttons button-row secondaryButtons">
             <div class="button-column" style="flex-grow: 1;-webkit-flex-grow: 1;">
               <div class="button-row">
```
Comment (zainafzal08) [inline js/foam/apps/calc/SecondaryButtonsView.js:31]: This changes the way navigation works as the old one relies on finding elements via aria labels in english, this update now makes the navigation work for all languages.

File: js/foam/graphics/ActionButtonCView.js [hunk 7]
```diff
@@ -210,7 +210,7 @@ CLASS({
 
       // Pressing space when has focus causes a synthetic press
       this.$.addEventListener('keypress', function(e) {
-        if ( ( e.charCode == 32 || e.charCode == '13')&& ! ( e.altKey || e.ctrlKey || e.shiftKey ) ) {
```
Comment (zainafzal08) [inline js/foam/graphics/ActionButtonCView.js:213]: Reserving the enter key for the equals operator makes more sense then my previous decision to allow it to trigger buttons

File: js/foam/ui/md/Halo.js [hunk 8]
```diff
@@ -199,30 +203,35 @@ CLASS({
   listeners: [
     {
       name: 'onMouseDown',
-      code: function(evt) {;
+      code: function(evt) {
```
Comment (zainafzal08) [inline js/foam/ui/md/Halo.js:206]: this makes it that clicking on a element does not show a visible focus UI but still switches the focus to the clicked element for future keyboard interactions.

File: js/foam/apps/calc/CalcView.js [hunk 9]
```diff
@@ -281,18 +284,18 @@ CLASS({
       justify-content: center;
       display: flex;
       align-items: center;
-      background-color: #4b4b4b;
+      background-color: #333333;
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: https://google.github.io/styleguide/htmlcssguide.html#Hexadecimal_Notation

File: js/foam/apps/calc/CalcView.js [hunk 10]
```diff
@@ -336,20 +339,20 @@ CLASS({
...
 
     .secondaryButtons .button {
-      background: rgb(52, 153, 128);
+      background: #00796b;
     }
 
     .tertiaryButtons {
       padding-left: 35px;
-      background: rgb(29, 233, 182);
+      background: #1DE9B6;
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: https://google.github.io/styleguide/htmlcssguide.html#Capitalization

File: js/foam/apps/calc/CalcView.js [hunk 11]
```diff
@@ -160,8 +161,10 @@ CLASS({
         const history = Array(historyNodeList.length).fill(0).map((_,i) => historyNodeList[i])
 
         // remove the equals speech label from first history elem
-        if (history[0])
-          history[0].setAttribute('aria-label', history[0].getAttribute('aria-label').replace(/^=/,''))
+        const equals = window.chrome.i18n ? window.chrome.i18n.getMessage('Calc_ActionSpeechLabel_equals') : 'equals';
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: Please clamp this to 80 chars.

File: js/foam/apps/calc/CalcView.js [hunk 12]
```diff
@@ -160,8 +161,10 @@ CLASS({
...
+
+        if (history[0] && history[0].getAttribute('aria-label'))
+          history[0].setAttribute('aria-label', history[0].getAttribute('aria-label').replace(new RegExp('^'+equals),''))
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: Please put this on multiple lines (80 chars), and I think it'd be easier to read with braces.

File: js/foam/apps/calc/Calc.js [hunk 13]
```diff
@@ -376,13 +395,25 @@ CLASS({
...
+          .getElementById('deg-label')
+          .setAttribute('aria-label', this.model_.DEG.label);
+      }
     },
     {
       name: 'rad',
       speechLabel: 'switch to radians',
       translationHint: 'short form for "radians" calculator mode',
-      code: function() { this.degreesMode = false; }
+      code: function() {
```
Comment (E-T-) [inline js/foam/apps/calc/Calc.js:410]: Arrow function?

File: js/foam/apps/calc/Calc.js [hunk 14]
```diff
@@ -376,13 +395,25 @@ CLASS({
       name: 'deg',
       speechLabel: 'switch to degrees',
       translationHint: 'short form for "degrees" calculator mode',
-      code: function() { this.degreesMode = true; }
+      code: function() {
```
Comment (E-T-) [inline js/foam/apps/calc/Calc.js:398]: arrow function?

File: js/foam/apps/calc/CalcView.js (hunk 12 above)
Comment (E-T-) [reply js/foam/apps/calc/CalcView.js]: See https://google.github.io/styleguide/jsguide.html#formatting-column-limit for context.

File: js/foam/graphics/AbstractCViewView.js [hunk 15]
```diff
@@ -160,7 +160,20 @@ CLASS({
       var tabIndex  = this.tabIndex ? ' tabindex="' + this.tabIndex + '"' : '';
       var arrowNav  = ''
 
-      const toSelector = (s) => (s && s[0] === '[') ? s.replace(/\[([^\[\]]+)\]/,'[aria-label=\'$1\'][aria-role=\'button\']') : s
+      function toSelector(s) {
+        if(s && s[0] === '[') {
```
Comment (E-T-) [inline js/foam/graphics/AbstractCViewView.js]: Add a space before the `(s &&...`

File: js/foam/graphics/AbstractCViewView.js [hunk 16]
```diff
@@ -160,7 +160,20 @@ CLASS({
...
+          let msg = null;
+          if(window.chrome && window.chrome.i18n) {
```
Comment (E-T-) [inline js/foam/graphics/AbstractCViewView.js]: Prepend a space to `(window.chrome...`

File: js/foam/graphics/AbstractCViewView.js [hunk 17]
```diff
@@ -160,7 +160,20 @@ CLASS({
...
+            msg = window.chrome.i18n.getMessage('Calc_ActionSpeechLabel_'+s.substr(1,s.length-2));
```
Comment (E-T-) [inline js/foam/graphics/AbstractCViewView.js]: 80 chars for line length

File: js/foam/graphics/AbstractCViewView.js [hunk 18]
```diff
@@ -160,7 +160,20 @@ CLASS({
...
+          }
+
+          if (msg) s ='['+msg+']';
+
+          return s.replace(/\[([^\[\]]+)\]/,'[aria-label=\'$1\'][aria-role=\'button\']');
```
Comment (E-T-) [inline js/foam/graphics/AbstractCViewView.js]: Is this essentially escaping stuff? Could you add a comment here? I feel like it's difficult to read or quickly grok the intent of this line.

File: js/foam/ui/md/Halo.js (hunk 8 above)
Comment (E-T-) [reply js/foam/ui/md/Halo.js:206]: Arrow function?

File: js/foam/apps/calc/CalcView.js [hunk 19]
```diff
@@ -160,8 +161,10 @@ CLASS({
         const history = Array(historyNodeList.length).fill(0).map((_,i) => historyNodeList[i])
 
         // remove the equals speech label from first history elem
-        if (history[0])
-          history[0].setAttribute('aria-label', history[0].getAttribute('aria-label').replace(/^=/,''))
+        const equals = window.chrome.i18n ? window.chrome.i18n.getMessage('Calc_ActionSpeechLabel_equals') : 'equals';
+
+        if (history[0] && history[0].getAttribute('aria-label'))
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: Could you help me understand what's happening here? Potentially a comment in code is appropriate, just trying to figure out what you're doing.

File: js/foam/ui/md/Halo.js (hunk 8 above)
Comment (zainafzal08) [reply js/foam/ui/md/Halo.js:206]: Do we want to deviate from the style of the rest of the codebase? I'm happy to use arrow functions but thought i rather as well match the way the current codebase works

File: js/foam/apps/calc/CalcView.js (hunk 9 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: Thank you!

File: js/foam/apps/calc/CalcView.js (hunk 10 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: Thank you!

File: js/foam/apps/calc/CalcView.js (hunk 11 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: Will do!

File: js/foam/apps/calc/CalcView.js (hunk 12 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: Thank you :D

File: js/foam/graphics/AbstractCViewView.js (hunk 18 above)
Comment (zainafzal08) [reply js/foam/graphics/AbstractCViewView.js]: Good point it is pretty hard to parse, will add a comment

File: js/foam/ui/md/Halo.js (hunk 8 above)
Comment (E-T-) [reply js/foam/ui/md/Halo.js:206]: I'll leave it to you to decide, but my preference is generally to use an arrow function and go for incremental improvements. I can be convinced by arguments of consistency but I have a preference for making code "better" (for subjective definition better) if possible.

Comment (zainafzal08) [reply js/foam/ui/md/Halo.js:206]: Fair nuff, will update!

File: js/foam/apps/calc/Calc.js (hunk 13 above)
Comment (zainafzal08) [reply js/foam/apps/calc/Calc.js:410]: Thanks!

File: js/foam/apps/calc/CalcView.js (hunk 19 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: I chucked in a comment :)

```
        The way the history is set up means that every element in the history
        has to be rendered the same way, as such every other number element has
        the aria label "equals <num>". This means as you are tabbing through
        something such as
         A
         + B
        --------
         C
        The screen reader helpfully announces "A + B equals C" as "C" has
        the "equals C" aria label. Sadly we don't want this label on A, only
        on C and any subsequent numbers. This bit of JS removes that label
        from the first element in the history.
```

File: js/foam/ui/md/Halo.js (hunk 8 above)
Comment (zainafzal08) [reply js/foam/ui/md/Halo.js:206]: turns out these functions get .bind called on them via the magic of foam and as such doing arrow functions breaks them. Back to anon functions we go!

File: js/foam/graphics/AbstractCViewView.js [hunk 20]
```diff
@@ -158,9 +158,40 @@ CLASS({
       var className = this.className ? ' class="' + this.className + '"' : '';
       var title     = this.speechLabel ? ' aria-role="button" aria-label="' + this.speechLabel + '"' : '';
       var tabIndex  = this.tabIndex ? ' tabindex="' + this.tabIndex + '"' : '';
-      var arrowNav  = ''
-
-      const toSelector = (s) => (s && s[0] === '[') ? s.replace(/\[([^\[\]]+)\]/,'[aria-label=\'$1\'][aria-role=\'button\']') : s
+      var arrowNav  = '';
+
+      // normally to target a button in the calculator display you would need
```
Comment (E-T-) [inline js/foam/graphics/AbstractCViewView.js]: Please comment with sentence case :) Add full stops etc. as necessary :)

File: js/foam/apps/calc/SecondaryButtonsView.js (hunk 6 above)
Comment (E-T-) [reply js/foam/apps/calc/SecondaryButtonsView.js:31]: Nice! :)

File: js/foam/apps/calc/CalcView.js [hunk 21]
```diff
@@ -159,9 +160,33 @@ CLASS({
...
+        //  + B
+        // --------
+        //  C
+        //
+        // The screen reader helpfully announces "A + B equals C" as "C" has
+        // the "equals C" aria label. Sadly we don't want this label on A, only
+        // on C and any subsequent numbers. This bit of JS removes that label
+        // from the first element in the history.
+
+        let equals = 'equals';
```
Comment (E-T-) [inline js/foam/apps/calc/CalcView.js]: I don't fully believe this is better but I thought I'd suggest it

`const equalsMessage = window.chrome.i18n && window.chrome.i18n.getMessage('Calc_ActionSpeechLabel_equals') || 'equals';`

**Review by E-T- — APPROVED** [review]
I think this is fine. Thanks for explaining why you can't use arrow functions! :)

File: js/foam/apps/calc/CalcView.js (hunk 21 above)
Comment (zainafzal08) [reply js/foam/apps/calc/CalcView.js]: seems neater to me! will update

**Review by dstoc — APPROVED** [review]
